spec/controllers/users_controller_spec.rb

```diff
@@ -120,34 +120,78 @@
           expect(response).to redirect_to(user)
         end
-      end
 
-      describe "with invalid params" do
-        it "assigns the user as @user" do
-          allow_any_instance_of(User).to receive(:save).and_return(false)
-          put :update, params: { id: user.to_param, user: { name: 'invalid value' } }
-          expect(assigns(:user)).to eq(user)
-        end
+        context "communication opt-in" do
+          before do
+            Timecop.freeze(Time.new(2002, 10, 31))
+          end
+
+          after do
+            Timecop.return
+          end
 
-        it "re-renders the 'edit' template" do
-          allow_any_instance_of(User).to receive(:save).and_return(false)
-          put :update, params: { id: user.to_param, user: { name: 'invalid value' } }
-          expect(response).to render_template("edit")
+          shared_examples_for 'tracks opt-in time' do |attribute|
+            it "sets #{attribute} to the current time" do
+              put :update, params: { id: user.to_param, user: { attribute => '1' } }
+              expect(user.reload.send "#{attribute}_at").to eq(Time.new(2002, 10, 31))
+            end
+          end
+
+          it_behaves_like 'tracks opt-in time', :opted_in_newsletter
+          it_behaves_like 'tracks opt-in time', :opted_in_announcements
+          it_behaves_like 'tracks opt-in time', :opted_in_marketing_announcements
+          it_behaves_like 'tracks opt-in time', :opted_in_surveys
+          it_behaves_like 'tracks opt-in time', :opted_in_sponsorships
+          it_behaves_like 'tracks opt-in time', :opted_in_applications_open
+
+          context 'opting out' do
+            shared_examples_for 'tracks opt-out' do |attribute|
+              before do
+                user.update_attribute attribute, Time.new(2002, 10, 31)
+              end
+
+              it "sets #{attribute} to nil" do
+                put :update, params: { id: user.to_param, user: { attribute => '0' } }
+                expect(user.reload.send "#{attribute}_at").to eq(nil)
+              end
+            end
+
+            it_behaves_like 'tracks opt-out', :opted_in_newsletter
+            it_behaves_like 'tracks opt-out', :opted_in_announcements
+            it_behaves_like 'tracks opt-out', :opted_in_marketing_announcements
+            it_behaves_like 'tracks opt-out', :opted_in_surveys
+            it_behaves_like 'tracks opt-out', :opted_in_sponsorships
+            it_behaves_like 'tracks opt-out', :opted_in_applications_open
+          end
         end
       end
+    end
 
-      context "another user's profile" do
-        let!(:another_user) { create(:user) }
+    describe "with invalid params" do
+      it "assigns the user as @user" do
+        allow_any_instance_of(User).to receive(:save).and_return(false)
+        put :update, params: { id: user.to_param, user: { name: 'invalid value' } }
+        expect(assigns(:user)).to eq(user)
```

Comment:
Testing against `assigns` is deprecated since Rails 5 and also no longer the way people want to test Rails applications 😿 Therefore, since the app will probably be maintained for some more time, let's not add new things and base them on already deprecated API.

However, I also don't understand the point of the test anyways I think 🙈 What do you want to test? Is there any **outcome** of the request that you're interested in? Maybe the page shows the name of the user, or the page shows the form again, etc? If so, please test this.
If this is to make sure that the controller assigns the right user: just remove the test and ensure this quality of the controller by another test that actually checks the outcome of a request 😉

Reply:
Whoa. No idea how this got in. Wonder if I rebased something wrong....